### PR TITLE
[RFC] Update Design documentation based on rav1e 0.4.0

### PR DESCRIPTION
**Pull request by vibhoothiiaanand**
This PR Updates the design section,

+ Adds new Structure.md which explains the design of rav1e 0.3.1

Would require to get more feedback on the verbosity of explanation and correctness.

Currently, design section was last updated ~2 Years and we did have nearly revamp of most of the file-structure but still have some of it.

As the file structure will be changed over-time, we would require to update it and cross-check on every major release (0.X.0), hopefully, I can volunteer to do that over-time.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -118,17 +118,7 @@ https://arewecompressedyet.com/analyzer/?d=https://people.xiph.org/~mbebenita/an
...
-* src/lib.rs - The top level library, contains code to write headers, manage buffers, and iterate through each superblock.
-* src/partition.rs - Functions and enums to manage partitions (subdivisions of a superblock).
-* src/predict.rs - Intra prediction implementations.
-* src/quantize.rs - Quantization and dequantization functions for coefficients.
-* src/rdo.rs - RDO-related structures and distortion computation functions.
-* src/transform/*.rs - Implementations of DCT and ADST transforms.
-* src/util.rs - Misc utility code.
-* src/bin/rav1e.rs - rav1e command line tool.
-* src/bin/rav1erepl.rs - Command line tool for debugging.
+The File Structure and design of the encoder is being explained more in [Structure](doc/STRUCTURE.md) docuement.
```
Comment (shssoichiro) [inline README.md:121]:
```suggestion
The File Structure and design of the encoder is explained more in [Structure](doc/STRUCTURE.md) document.
```

File: doc/STRUCTURE.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
+
+# File Structure of the rav1e 0.4.0
```
Comment (shssoichiro) [inline doc/STRUCTURE.md:2]:
```suggestion
# File Structure of rav1e 0.4.0
```

File: doc/STRUCTURE.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+
+## High-level directory structure
+
+![Image](https://i.imgur.com/UiFieT5.png)
```
Comment (shssoichiro) [inline doc/STRUCTURE.md:7]: I'd prefer to have this image hosted in the git repo. I think that's possible with markdown in Github.

File: doc/STRUCTURE.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+| [arm/32/*.S](../src/arm/32/)                          | ARM optimised functions for different encoding tools imported from dav1d by release                        |
+| [arm/64/*.S](../src/arm/64)                           | AArch64 optimised functions for different encoding tools imported from dav1d by release                    |
+| [arm/asm.S](../src/arm/asm.S)                         | Common functions used for Assembly implementation                                                          |
+| [arm/table.S](../src/arm/table.S)                     | Tables for various ARM optimised functions                                                                 |
+| [asm/\*/*.rs](../src/asm/)                            | High-level functions for binding rust and assembly functions for x86 and AArch64 Architecture              |
+| [bin/common.rs](../src/bin/common.rs)                 | Functions, enums, structures used command-line tool and debugging                                          |
+| [bin/rav1e.rs](../src/bin/rav1.rs)                    | High-level functions for encoding and decoding using y4m decoder                                           |
```
Comment (shssoichiro) [inline doc/STRUCTURE.md:24]:
```suggestion
| [bin/rav1e.rs](../src/bin/rav1e.rs)                    | CLI Interface for encoding from y4m files with rav1e                                           |
```

File: doc/STRUCTURE.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+| [bin/stats.rs](../src/bin/stats.rs)                   | Functions for displaying Frame summary, progress info, metrics of the encoding process                     |
+| [bin/kv.rs](../src/bin/kv.rs)                         | Serialisation configuration of Key-value strings                                                           |
+| [bin/errror.rs](../src/bin/error.rs)                  | Functions and enums to parse various errors and displaying                                                 |
+| [bin/muxer/*.rs](../src/bin/muxer/)                   | Contains IVF Muxer functions for header definition, writing frames and flushing                            |
+| [bin/decoder/*.rs](../src/bin/decoder/)               | Decoder related structures and functions                                                                   |
+| [capi.rs](../src/capi.rs)                             | C Compatible API for rav1e for using as a library                                                          |
```
Comment (shssoichiro) [inline doc/STRUCTURE.md:30]:
```suggestion
| [capi.rs](../src/capi.rs)                             | C Compatible API for using rav1e as a library                                                          |
```

File: doc/STRUCTURE.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+| [header.rs](../src/header.rs)                         | The enums and structs of bitstream headers for writing                                                     |
+| [lib.rs](../src/lib.rs)                               | The top level library, contains code to write headers, manage buffers, and iterate through each superblock |
+| [lrf.rs](../src/lrf.rs)                               | Low-level implementation of Loop restoration filter                                                        |
+| [mc.rs](../src/mc.rs)                                 | Low-level implementation of Motion Compensation of the encoding process                                    |
+| [me.rs](../src/me.rs)                                 | Motion Estimation related structures and functions of the encoding process                                 |
+| [partition.rs](../src/partition.rs)                   | Functions and enums to manage partitions (subdivisions of a superblock)                                    |
+| [predict.rs](../src/predict.rs)                       | Intra prediction implementations                                                                           |
```
Comment (shssoichiro) [inline doc/STRUCTURE.md:48]:
```suggestion
| [predict.rs](../src/predict.rs)                       | Intra and inter prediction implementations                                                                           |
```

File: README.md [hunk 7]
```diff
@@ -118,17 +118,7 @@ https://arewecompressedyet.com/analyzer/?d=https://people.xiph.org/~mbebenita/an
...
-* src/lib.rs - The top level library, contains code to write headers, manage buffers, and iterate through each superblock.
-* src/partition.rs - Functions and enums to manage partitions (subdivisions of a superblock).
-* src/predict.rs - Intra prediction implementations.
-* src/quantize.rs - Quantization and dequantization functions for coefficients.
-* src/rdo.rs - RDO-related structures and distortion computation functions.
-* src/transform/*.rs - Implementations of DCT and ADST transforms.
-* src/util.rs - Misc utility code.
-* src/bin/rav1e.rs - rav1e command line tool.
-* src/bin/rav1erepl.rs - Command line tool for debugging.
+The File Structure and design of the encoder is explained more in [Structure](doc/STRUCTURE.md) document.
```
Comment (lu-zero) [inline README.md:121]:
```suggestion
The File Structure and design of the encoder is explained more in the [Structure](doc/STRUCTURE.md) document.
```

**Comment by lu-zero** [comment]
Please rebase, fix the nit and land it. :)

**Pull request merged**